alexandria/core/validations.py

```diff
@@ -62,17 +64,40 @@ def validate_file(self, data, context):
         validate_file_infection(data["content"])
 
         # Validate that the mime type is allowed in the category
-        mime_type = data["content"].content_type
-        if mime_type == "application/octet-stream" or not mime_type:
-            guess, encoding = guess_type(data["name"])
-            if guess is not None:
-                mime_type = guess
-            else:
-                data["content"].seek(0)
-                mime_type = magic.from_buffer(data["content"].read(), mime=True)
-                data["content"].seek(0)
-
-        validate_mime_type(mime_type, data["document"].category)
-        data["mime_type"] = mime_type
+        content_type_header = data["content"].content_type
+        extension_type, _ = guess_type(data["name"])
+
+        if not content_type_header:  # pragma: no cover
+            raise ValidationError(gettext_lazy("Content-Type header is missing."))
+        if not extension_type:
+            raise ValidationError(gettext_lazy("The uploaded file has no extension."))
+
+        if content_type_header == "application/octet-stream":
+            content_type_header = extension_type
+        if content_type_header != extension_type:
+            raise ValidationError(
+                gettext_lazy(
+                    "Content type %(content_type)s does not match file extension %(extension)s."
+                    % {"content_type": content_type_header, "extension": extension_type}
+                )
+            )
+
+        data["content"].seek(0)
+        file_content_type = magic.from_buffer(data["content"].read(), mime=True)
+        data["content"].seek(0)
+
+        if file_content_type != content_type_header:
+            raise ValidationError(
+                gettext_lazy(
+                    "Content type %(content_type)s does not match detected file content %(file_content_type)s."
+                    % {
+                        "content_type": content_type_header,
+                        "file_content_type": file_content_type,
+                    }
+                )
+            )
```

Comment:
I sort of like it, but also see some problems: If the upload for some reason mismatches the mime types, the user gets an error that they can't really do anything about.

Reply:
as discussed, we might have to "loosen" this up if it turns out to be too restrictive in practice.